google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -1710,11 +1785,20 @@ private MultiplexedSession getOrNull() {
     @Override
     public MultiplexedSession get() {
       try {
-        MultiplexedSession multiplexedSession = super.get();
-        if (multiplexedSession != null) {
-          multiplexedSession.markBusy(span);
-          span.addAnnotation("Using Session", "sessionId", multiplexedSession.getName());
-          incrementNumSessionsInUse(true);
+        if (multiplexedSession == null) {
+          // access the settableApiFuture object before acquiring lock to avoid deadlock
+          SessionImpl sessionImpl =
+              new SessionImpl(
+                  sessionClient.getSpanner(), currentMultiplexedSessionReference.get().get());
+          synchronized (lock) {
```

Comment:
nit: I think that you only need to take the lock for the `incrementNumSessionsInuse(true);` call. The rest does not need the session pool lock.

Reply:
Yeah, I removed the lock from here since we are acquiring the lock again within the `incrementNumSessionsInUse` method.